manta-trusted-setup/src/groth16/kzg.rs

```diff
@@ -0,0 +1,784 @@
+            + C::uncompressed_size(&self.tau.ratio.1)
+            + C::uncompressed_size(&self.alpha.ratio.0)
+            + C::uncompressed_size(&self.alpha.ratio.1)
+            + C::uncompressed_size(&self.beta.ratio.0)
+            + C::uncompressed_size(&self.beta.ratio.1)
+            + C::uncompressed_size(&self.tau.matching_point)
+            + C::uncompressed_size(&self.alpha.matching_point)
+            + C::uncompressed_size(&self.beta.matching_point)
+    }
+}
+
+impl<C> CanonicalDeserialize for Proof<C>
+where
+    C: Deserializer<C::G1> + Deserializer<C::G2> + Pairing,
+{
+    #[inline]
+    fn deserialize<R>(mut reader: R) -> Result<Self, SerializationError>
+    where
+        R: Read,
+    {
+        let tau_ratio = (
+            NonZero::<C>::deserialize_compressed(&mut reader)?,
+            NonZero::<C>::deserialize_compressed(&mut reader)?,
+        );
+        let alpha_ratio = (
+            NonZero::<C>::deserialize_compressed(&mut reader)?,
+            NonZero::<C>::deserialize_compressed(&mut reader)?,
+        );
+        let beta_ratio = (
+            NonZero::<C>::deserialize_compressed(&mut reader)?,
+            NonZero::<C>::deserialize_compressed(&mut reader)?,
+        );
+        Ok(Self {
+            tau: RatioProof {
+                ratio: tau_ratio,
+                matching_point: NonZero::<C>::deserialize_compressed(&mut reader)?,
+            },
+            alpha: RatioProof {
+                ratio: alpha_ratio,
+                matching_point: NonZero::<C>::deserialize_compressed(&mut reader)?,
+            },
+            beta: RatioProof {
+                ratio: beta_ratio,
+                matching_point: NonZero::<C>::deserialize_compressed(&mut reader)?,
+            },
+        })
+    }
+
+    #[inline]
+    fn deserialize_uncompressed<R>(mut reader: R) -> Result<Self, SerializationError>
+    where
+        R: Read,
+    {
+        let tau_ratio = (
+            NonZero::<C>::deserialize_uncompressed(&mut reader)?,
+            NonZero::<C>::deserialize_uncompressed(&mut reader)?,
+        );
+        let alpha_ratio = (
+            NonZero::<C>::deserialize_uncompressed(&mut reader)?,
+            NonZero::<C>::deserialize_uncompressed(&mut reader)?,
+        );
+        let beta_ratio = (
+            NonZero::<C>::deserialize_uncompressed(&mut reader)?,
+            NonZero::<C>::deserialize_uncompressed(&mut reader)?,
+        );
+        Ok(Self {
+            tau: RatioProof {
+                ratio: tau_ratio,
+                matching_point: NonZero::<C>::deserialize_uncompressed(&mut reader)?,
+            },
+            alpha: RatioProof {
+                ratio: alpha_ratio,
+                matching_point: NonZero::<C>::deserialize_uncompressed(&mut reader)?,
+            },
+            beta: RatioProof {
+                ratio: beta_ratio,
+                matching_point: NonZero::<C>::deserialize_uncompressed(&mut reader)?,
+            },
+        })
+    }
+}
```

Comment:
Unfortunately, to conform to the ZCash serialization format we have to do this in a non-standard way, otherwise yes we would just use `CanonicalSerialize` for everything.